pkg/operator/ceph/pool/radosnamespace/controller.go

```diff
@@ -0,0 +1,312 @@
+		logger.Debug("external rados namespace creation is not supported, create it manually, the controller will assume it's there")
+	} else {
+		// Create or Update rados namespace
+		err = r.createOrUpdateRadosNamespace(cephBlockPoolRadosNamespace)
+		if err != nil {
+			if strings.Contains(err.Error(), opcontroller.UninitializedCephConfigError) {
+				logger.Info(opcontroller.OperatorNotInitializedMessage)
+				return opcontroller.WaitForRequeueIfOperatorNotInitialized, nil
+			}
+			r.updateStatus(r.client, request.NamespacedName, cephv1.ConditionFailure)
+			return reconcile.Result{}, errors.Wrapf(err, "failed to create or update ceph pool rados namespace %q", cephBlockPoolRadosNamespace.Name)
+		}
+	}
+	// Update CSI config map
+	// If the mon endpoints change, the mon health check go routine will take care of updating the
+	// config map, so no special care is needed in this controller
+	csiClusterConfigEntry := csi.CsiClusterConfigEntry{
+		Monitors:       csi.MonEndpoints(r.clusterInfo.Monitors),
+		RadosNamespace: cephBlockPoolRadosNamespace.Name,
+	}
+	err = csi.SaveClusterConfig(r.context.Clientset, buildClusterID(cephBlockPoolRadosNamespace), r.clusterInfo, &csiClusterConfigEntry)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "failed to save cluster config")
+	}
+	// Success! Let's update the status
+	if cephCluster.Spec.External.Enable {
+		r.updateStatus(r.client, request.NamespacedName, cephv1.ConditionConnected)
+	} else {
+		r.updateStatus(r.client, request.NamespacedName, cephv1.ConditionReady)
+	}
+
+	// Return and do not requeue
+	logger.Debug("done reconciling")
+	return reconcile.Result{}, nil
+}
+
+// Create the ceph blockpool rados namespace
+func (r *ReconcileCephBlockPoolRadosNamespace) createOrUpdateRadosNamespace(cephBlockPoolRadosNamespace *cephv1.CephBlockPoolRadosNamespace) error {
+	logger.Infof("creating ceph blockpool rados namespace %s in namespace %s", cephBlockPoolRadosNamespace.Name, cephBlockPoolRadosNamespace.Namespace)
+
+	err := cephclient.CreateRadosNamespace(r.context, r.clusterInfo, cephBlockPoolRadosNamespace.Spec.BlockPoolName, cephBlockPoolRadosNamespace.Name)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create ceph blockpool rados namespace %q", cephBlockPoolRadosNamespace.Name)
+	}
+
+	return nil
+}
+
+// Delete the ceph blockpool rados namespace
+func (r *ReconcileCephBlockPoolRadosNamespace) deleteRadosNamespace(cephBlockPoolRadosNamespace *cephv1.CephBlockPoolRadosNamespace) error {
+	logger.Infof("deleting ceph blockpool rados namespace object %q", cephBlockPoolRadosNamespace.Name)
+	if err := cephclient.DeleteRadosNamespace(r.context, r.clusterInfo, cephBlockPoolRadosNamespace.Spec.BlockPoolName, cephBlockPoolRadosNamespace.Name); err != nil {
+		return errors.Wrapf(err, "failed to delete ceph blockpool rados namespace %q", cephBlockPoolRadosNamespace.Name)
+	}
+
+	logger.Infof("deleted ceph blockpool rados namespace %q", cephBlockPoolRadosNamespace.Name)
```

Comment:
The rook operator can operate on multiple resources in multiple namespaces. All info/debug messages must list the namespaced name so context is clear.